src/ripple/shamap/SHAMapTreeNode.h

```diff
@@ -172,11 +177,47 @@ class SHAMapInnerNode
     void updateHashDeep();
     void addRaw (Serializer&, SHANodeFormat format) const override;
     std::string getString (SHAMapNodeID const&) const override;
+    uint256 const& key() const override;
+    void invariants(bool is_v2, bool is_root = false) const override;
 
     friend std::shared_ptr<SHAMapAbstractNode>
         SHAMapAbstractNode::make(Blob const& rawNode, std::uint32_t seq,
              SHANodeFormat format, SHAMapHash const& hash, bool hashValid,
-                 beast::Journal j);
+                 beast::Journal j, SHAMapNodeID const& id);
+
+    friend class SHAMapInnerNodeV2;
+};
+
+class SHAMapTreeNode;
+
+// SHAMapInnerNodeV2 is a "version 2" inner node.  It always has at least two children
+//   unless it is the root node and there is only one leaf in the tree, in which case
+//   that leaf is a direct child of the root.
+class SHAMapInnerNodeV2
+    : public SHAMapInnerNode
```

Comment:
WHOA.... derivation and virtual functions? Isn't it the case that a SHAMap has either all version 1 or all version 2 inner nodes? It seems we're paying the cost of the run-time choice when we already know the version of all possible inner nodes the moment the SHAMap is constructed. Or am I missing something?